# Calculates Differential Gene Expression (DGE)

Code in this folder is used to define which genes are Differentially Expressed.
These files expect gene expression counts, generated by command lines at the end of this instruction set.


Instructions:

1. Run `calc_diff_gene_exp.R` to compute gene expression levels (logFC, FDR and p-value) for all genes in HS and MM.

Then, two pipelines can be chosen: `conserved` or `pooling`.

2. Run `select_dge_genes-<pipeline>.py` to select only those genes that are Differentialy Expressed (DE) based on assumed parameters (ex.: FDR, logFC & CPM).


Other files:
- `calc_plot_HS_PCA.py` and `calc_plot_MM_PCA.py` calculate Principal Component Analysis (PCA) and plot the results.
- `stats_genes.py` generates statistis about the DGE genes, the output of which is shown in `stats_genes.md`.


## R Session Info

```R
> sessionInfo()
R version 3.6.1 (2019-07-05)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_3.2.1 edgeR_3.26.7  limma_3.40.6 

loaded via a namespace (and not attached):
 [1] locfit_1.5-9.1   Rcpp_1.0.2       lattice_0.20-38  withr_2.1.2      assertthat_0.2.1
 [6] dplyr_0.8.3      crayon_1.3.4     R6_2.4.0         grid_3.6.1       gtable_0.3.0    
[11] magrittr_1.5     scales_1.0.0     pillar_1.4.2     rlang_0.4.0      lazyeval_0.2.2  
[16] rstudioapi_0.10  tools_3.6.1      glue_1.3.1       purrr_0.3.2      munsell_0.5.0   
[21] compiler_3.6.1   pkgconfig_2.0.2  colorspace_1.4-1 tidyselect_0.2.5 tibble_2.1.3   
```

---

# Calculates FPKM

We also calculated FPKM values for Spermatocytes (middle). These are located in files `data/DM/DM_Spermatocytes_FPKM_sample1.csv` and `data/DM/DM_Spermatocytes_FPKM_sample2.csv`
After read aligment with `Hisat2` the following command line was used to calculate values.

```bash
stringtie DM_aligned_reads.bam -G DM_annotation.gtf -e -B –A DM_Sample.txt –o Sample_rep.gtf
```

---

## Original RNAseq data


Homo Sapiens:
- [SRP069329](https://www.ncbi.nlm.nih.gov/sra/?term=SRP069329)

> Jan, et al. *Development* 2017 144: 3659-3673; doi: [10.1242/dev.152413](https://dev.biologists.org/content/144/20/3659)


Mus Musculus:
- [GSE43717](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE43717)

> Soumillon, et al. *Cell Rep*. 2013 Jun 27;3(6):2179-90. doi: [10.1016/j.celrep.2013.05.031](https://www.sciencedirect.com/science/article/pii/S2211124713002489)


Drosophila Melanogaster:
- [PRJNA496287](https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA496287)

> Vedelek, et al. *BMC Genomics*. 2018 Sept 24;19(1):697. doi: [10.1186/s12864-018-5085-z](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-018-5085-z)

## Original MicroArray data

Drosophila Melanogaster
- [GSE18502](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE18502)

> Vibranovski, et al *PLOS Genetics* 2009 5(11): e1000731; doi: [10.1371/journal.pgen.1000731](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1000731)

---

## From FASTQ to count data

Command line tools used to tranform the FASTQ files into feature count matrices.

Homo Sapiens:
- Alignment to Human Genome
```bash
hisat2 -x HS_index -U HS_Dataset_1_reads.fq.gz | samtools view -b | samtools sort -o HS_Dataset_1.bam
```

- Feature counts
```bash
featureCounts -a HS_annotation.gtf -t exon -g gene_id -o HS_Dataset_1.txt HS_Dataset_1.bam
```

Mouse:
- Alignment to Mouse Genome
```bash
hisat2 --rna-strandness F -x MM_index -U MM_Dataset_1_reads.fq.gz | samtools view -b | samtools sort -o MM_Dataset_1.bam
```
- Feature counts
```bash
featureCounts –s1 -a MM_annotation.gtf -t exon -g gene_id -o MM_Dataset_1.txt MM_Dataset_1.bam
```
